### DISCUSSION TITLE
Support for AWS Lambda Streaming Responses in `@remix-run/architect`

### DISCUSSION THREAD

**Discussion by prxgiusti** [remix-run/remix#6037] [Proposals]
AWS recently announced [support for streaming responses](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-response-streaming/) in AWS Lambda. It seems that this will require modification of the `createRequestHandler` function in `@remix-run/architect` to properly receive the Lambda's `responseStream` and pass it along to `sendRemixResponse`, which will then pipe it into the Remix response.

Thank you for considering this feature request.

**Comment by prxgiusti** [comment]
It's important to note that the API Gateway won't send responses via stream and that it will be necessary to directly activate the lambda through its URL (or make that URL an origin in Cloudfront). Consequently, `createRequestHandler` could no longer return the type `APIGatewayProxyHandlerV2` nor receive events only from the APIGateway.

Perhaps the best solution would be to create another function that can be used by applications that will use response via stream (`createStreamRequestHandler`).

**Comment by prxgiusti** [comment]
If you're seeking a solution, I've managed to address the issue by making modifications to the [remix-architecture](https://github.com/remix-run/remix/blob/main/packages/remix-architect/server.ts) `server.ts` file. I've shared my revised version in [this repository](https://github.com/prxg22/poc-frontend-infra/blob/main/server/utils/remix/aws.ts). And that's how you use the helper:

```tsx
import * as build from '@remix-run/dev/server-build'
import { installGlobals } from '@remix-run/node'

import { createStreamRequestHandler } from './utils/remix/aws'
installGlobals()

export const handler = createStreamRequestHandler({
  build,
  mode: process.env.NODE_ENV,
})
```



It's important to keep in mind that when working with APIGateway, there's a critical aspect to consider: the buffering of stream responses. Essentially, the response is only delivered once the entire stream is loaded. A different approach to achieve stream responses is creating a [Lambda function URL](https://docs.aws.amazon.com/lambda/latest/dg/lambda-urls.html).

**Reply by benjamindulau** [reply]
@prxg22 Thanks to your code I've managed to make `RESPONSE_STREAM` mode works, but somewhat AWS still seems to buffer the entire response before sending it to the client. Have you got an actual defer/streaming working?

**Reply by prxgiusti** [reply]
Yes, I did using the handler I mentioned. Check if you configured your lamba and function url correctly.

**Reply by benjamindulau** [reply]
@prxg22 Ok. So you get the first chunk from the stream immediately? There is something weird for me, if I run the following handler:
```typescript
export const handler = awslambda.streamifyResponse(
  async (
      event: APIGatewayProxyEventV2, responseStream: Writable & {
          setContentType: (type: string) => void
      },
  ) => {
    responseStream.setContentType("text/plain");

    responseStream.write("Hello World");
    responseStream.write(" ");

    await new Promise((r) => setTimeout(r, 1000));

    responseStream.write("Goodbye");
    responseStream.end();
  },
);
``` 

`Hello World` is sent after 1sec instead of immediatly. I checked 10 times my configuration. The lambda is correctly configured to `RESPONSE_STREAM` invokation mode.

**Reply by prxgiusti** [reply]
Are you accessing your lambda directly via [Lambda Url](https://docs.aws.amazon.com/lambda/latest/dg/lambda-urls.html)? 
You can't stream responses through API Gateway, even if your lambda is configured to `RESPONSE_STREAM`, the API Gateway will buffer the whole response before send it.

**Reply by benjamindulau** [reply]
@prxg22 Yes. In fact streaming works. What I discovered is that for some reasons Lambda doesn't send the first chunk immediately into the stream. So I was experiencing things like all data from deferred promises to be streamed at the same moment.

I have exactly the behavior described here https://betterdev.blog/lambda-response-streaming-flush-content/

I'm forced to hack every loader payloads to add a big empty string so Lambda sends the chunk immediately... ugly :D

**Reply by owlyowl** [reply]
@prxg22 Hi Paulo
https://github.com/prxg22/poc-frontend-infra/blob/main/server/utils/remix/aws.ts I'm not sure how your example would work as it's still using API gateway rather than a direct invoke.

Did you modify it after you wrote the original? It still seems to be dealing with API gateway events

**Reply by prxgiusti** [reply]
> @prxg22 Hi Paulo
> https://github.com/prxg22/poc-frontend-infra/blob/main/server/utils/remix/aws.ts I'm not sure how your example would work as it's still using API gateway rather than a direct invoke.
> 
> Did you modify it after you wrote the original? It still seems to be dealing with API gateway events

No,@owlyowl, I'm using function url and CDN to access the lambda. [Here's](https://github.com/prxg22/poc-frontend-infra/blob/main/index.ts#L71) where I create the function url (I'm using pulumi to deploy the application). 
You must be misleadead by the readme, ignore it.

**Reply by prxgiusti** [reply]
> @prxg22 Yes. In fact streaming works. What I discovered is that for some reasons Lambda doesn't send the first chunk immediately into the stream. So I was experiencing things like all data from deferred promises to be streamed at the same moment.
> 
> I have exactly the behavior described here https://betterdev.blog/lambda-response-streaming-flush-content/
> 
> I'm forced to hack every loader payloads to add a big empty string so Lambda sends the chunk immediately... ugly :D
> 

@benjamindulau it's been a while since the last time I did this POC. I dont remember expericing this behavior, my example streams a list of many items from an API, maybe it always respond with enough size that I could not reproduce the behavior.

**Comment by owlyowl** [comment]
Thanks Paulo
I was a bit confused because your incoming server types are all still api
gateway related.

**Reply by prxgiusti** [reply]
@owlyowl I see. There's no specific types for function url events,  but it follows the API Gateway specification.

**Reply by owlyowl** [reply]
Thanks so much for replying I really appreciate it! Time to test this out with a few defer suspense calls :)

**Comment by owlyowl** [comment]
Thanks it does. I managed to run it up. Looks good and streaming works as
expected :)

**Reply by gitaugakwa** [reply]
I'm trying to get it to work, could be cause I'm on remix v2. I'm really not too sure how to get it working on aws

**Comment by owlyowl** [comment]
We found that `defer` is quite broken until React 18.3 so until that
resolves it makes SSR not as compelling.
API Gateway doesn't support streaming lambdas, you'll need to remove API
gateway and make the lambda direct invoke in order to enable streaming.

AWS have mentioned streaming behind API Gateway is on the roadmap.

**Comment by godu** [comment]
I have some troubles with defer. The first chunk is delayed.
Maybe this issue could help https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/issues/94

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
